source/dub/dub.d

```diff
@@ -757,12 +757,24 @@ class Dub {
 			return m_packageManager.getPackage(packageId, ver, dstpath);
 		}
 
-		auto path = getTempFile(packageId, ".zip");
-		supplier.fetchPackage(path, packageId, dep, (options & FetchOptions.usePrerelease) != 0); // Q: continue on fail?
-		scope(exit) std.file.remove(path.toNativeString());
-
-		logDiagnostic("Placing to %s...", placement.toNativeString());
-		return m_packageManager.storeFetchedPackage(path, pinfo, dstpath);
+		foreach_reverse (i; 0..4)
+		{
+			import std.zip : ZipException;
+			try {
+				auto path = getTempFile(packageId, ".zip");
+				supplier.fetchPackage(path, packageId, dep, (options & FetchOptions.usePrerelease) != 0); // Q: continue on fail?
+				scope(exit) std.file.remove(path.toNativeString());
+
+				logDiagnostic("Placing to %s...", placement.toNativeString());
+				return m_packageManager.storeFetchedPackage(path, pinfo, dstpath);
+			} catch (ZipException e) {
+				logInfo("Failed to extract zip archive for %s %s...", packageId, ver);
```

Comment:
This isn't the best behavior as the PackageManager fallbacks will never be tried if the main registry returns invalid zips. However, this would require refactoring their interface a bit and I wasn't sure whether this is wanted. In any case, this should already be an measureable improvement (see the test) as it will restart the download when the registry / GH returned invalid data.

Reply:
Packages are downloaded directly from GH though, at least that should be the case.
https://github.com/dlang/dub-registry/blob/fe18df23b32f602a7579816a801a8a4e98346b5c/source/dubregistry/web.d#L177-L187